Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+  rw [svdU₂, toColumns₂]
+  simp only [reindex_apply, Equiv.refl_symm, Equiv.coe_refl, submatrix_apply, id_eq]
+  funext i j
+  simp_rw [Matrix.mul_apply, conjTranspose_apply, of_apply,
+    ← conjTranspose_apply, IsHermitian.conjTranspose_eigenvectorMatrix,
+    ← Matrix.mul_apply, Matrix.mul_eq_one_comm.1 (IsHermitian.eigenvectorMatrix_mul_inv _)]
+  by_cases hij: i = j
+  · simp_rw [hij, one_apply_eq]
+  · rw [one_apply_ne hij, one_apply_ne]
+    simpa only [ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inr.injEq]
+
+lemma U₁'_conjTranspose_mul_U₂ (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A.svdU₁'ᴴ  *  A.svdU₂ = 0 := by
+  simp_rw [svdU₁', svdU₂, toColumns₁, toColumns₂, reindex_apply, Equiv.refl_symm, Equiv.coe_refl,
+    submatrix_apply, id_eq, HMul.hMul, dotProduct, conjTranspose_apply, of_apply,
+    ← conjTranspose_apply, IsHermitian.conjTranspose_eigenvectorMatrix, ← mul_apply,
+    Matrix.mul_eq_one_comm.1 (IsHermitian.eigenvectorMatrix_mul_inv _)]
+  funext i j
+  simp only [ne_eq, EmbeddingLike.apply_eq_iff_eq, not_false_eq_true, one_apply_ne, zero_apply]
+
+lemma mul_V₂_eq_zero (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A  *  A.svdV₂ = 0 := by
+  suffices h : Aᴴ * A * A.svdV₂ = 0
+  · exact (conjTranspose_mul_self_mul_eq_zero _ _).1 h
+  rw [reduced_spectral_theorem, Matrix.mul_assoc, V₁_conjTranspose_mul_V₂, Matrix.mul_zero]
+
+lemma conjTranspose_mul_U₂_eq_zero (A : Matrix (Fin M) (Fin N) 𝕂) : Aᴴ  *  A.svdU₂ = 0 := by
+  suffices h : A * Aᴴ * A.svdU₂ = 0
+  · exact (self_mul_conjTranspose_mul_eq_zero _ _).1 h
+  rw [reduced_spectral_theorem', Matrix.mul_assoc, U₁'_conjTranspose_mul_U₂]
+  simp only [Matrix.mul_zero]
+
+lemma U₁_conjTranspose_mul_U₂ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdU₁ᴴ  *  A.svdU₂ = 0 := by
+  unfold svdU₁
+  simp_rw [conjTranspose_mul, Matrix.mul_assoc, conjTranspose_mul_U₂_eq_zero, Matrix.mul_zero]
+
+lemma U₂_conjTranspose_mul_U₁ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdU₂ᴴ  *  A.svdU₁ = 0 := by
+  rw [← conjTranspose_conjTranspose (A.svdU₁), ← conjTranspose_mul, U₁_conjTranspose_mul_U₂,
+    conjTranspose_zero]
+
+lemma U_inv (A : Matrix (Fin M) (Fin N) 𝕂) :
+  (fromColumns A.svdU₁ A.svdU₂)ᴴ * (fromColumns A.svdU₁ A.svdU₂) = 1 := by
+  rw [conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromRows_mul_fromColumns,
+    U₁_conjTranspose_mul_U₂, U₁_conjTranspose_mul_U₁, U₂_conjTranspose_mul_U₂,
+    U₂_conjTranspose_mul_U₁, fromBlocks_one]
+
+lemma V_conjTranspose_mul_inj (A : Matrix (Fin M) (Fin N) 𝕂) {m : Type} :
+    Function.Injective (fun x : Matrix m (Fin N) 𝕂 => x  *  (fromColumns A.svdV₁ A.svdV₂)) := by
+  intro X Y h
+  replace h := congr_arg (fun x => x * (fromColumns A.svdV₁ A.svdV₂)ᴴ) h
+  dsimp at h
+  have V_inv' := V_inv A
+  rw [conjTranspose_fromColumns_eq_fromRows_conjTranspose, ← fromColumns_mul_fromRows_eq_one_comm,
+    ← conjTranspose_fromColumns_eq_fromRows_conjTranspose] at V_inv'
+  rwa [Matrix.mul_assoc, Matrix.mul_assoc, V_inv', Matrix.mul_one, Matrix.mul_one] at h
+  apply eigenColumnEquiv
```

Comment:
This is true for much more general reasons: if the matrix is square we could directly use [`IsUnit.mul_right_injective`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Units.html#IsUnit.mul_right_injective). So: could you prove the general statement `Matrix.mul_right_injective_of_inv` saying that `A * B = 1` implies the map `fun M => M * A` is injective, and the same on the left?